File: code/pa5/List.c
#include <stdlib.h>
#include "List.h"

typedef struct NodeObj*Node;

struct NodeObj{
		int value;
		Node next;
		Node previous;
}NodeObj;

	typedef struct ListObj{
		Node cursor;
		Node front;
		Node back;
		int length;
		int index;
	}ListObj;

	List newList(){
		List x = malloc (sizeof(struct ListObj));
		x->cursor=NULL;
        x->front=NULL;
        x->back=NULL;
        x->length=0;
        x->index=-1;
		return x;
	}


    int index(List L){
        if(L==NULL){
            return -1;
        }
        return L->index;
    }

    int length(List L){
        if(L==NULL){
            return -1;
        }
        return L->length;
    }

    int front(List L){
        if(L==NULL){
            printf("Null pointer error");
            return -1;
        }
    	if(L->length > 0){
    		return L->front->value;
    	}
        return -1;
    }

    int back(List L){
        if(L==NULL){
            printf("Null pointer error");
            return -1;
        }
    	if(L->length > 0){
    		return L->back->value;
    	}
        return -1;
    }

/*    int front(List N){
    	if(N-length > 0){
    		return N->length->value;
    	}
    	return -1;
    } */

    int get (List L){
    	if(L->length>0 && L->index >= 0){
    		return L->cursor->value;
    	}
    	return -1;
    }

    int equals(List A, List B){
    	Node x1 = A->front;
    	Node x2 = B->front;
    	while(x1!=NULL && x2!=NULL){
    		if(x1->value != x2->value){
    			return 0;
    		}
    		x1=x1->next;
    		x2=x2->next;
    	}
    	return 1;
    }


    void moveFront(List L){
    	if(L->front != NULL){
    		L->cursor = L->front;
    		L->index = 0;
    	}
    }

    void moveBack(List L){
        if(L->back != NULL){
            L->cursor = L->back;
            L->index = L->length-1;
        }
        
   }

    void movePrev(List L){
    	if(L->index > 0 && L->index <= L->length-1){
    		L->cursor = L->cursor->previous;
    		L->index--;
    	}else{
    		L->cursor=NULL;
            L->index = -1;
    	}

    }

    void moveNext(List L){
      if (L == NULL) {
        printf("List.c: 179: passing null ptr to moveNext");
        return;
    }

    if(L->index >= 0 && L->index < L->length - 1){
        L->cursor = L->cursor->next;
        L->index++;
    } else {
        L->cursor = NULL;
        L->index = -1;
    }
    }
    // void moveNext(List L){
    // 	if(L->index >0 && L->index <= L->length-1){
    // 		if(L->cursor != L->back){
    // 			L->cursor = L->cursor->next;
    // 			L->index++;
    // 		}else{
    // 			L->index = -1;
    // 			L->cursor = NULL;
    // 		}
    // 	}
    // 	printf("EMPTY LIST ERR");
    // }

    void prepend (List L, int x){
    	Node n = malloc(sizeof(struct NodeObj));
    	n->value = x;
        n->previous = NULL;
        n->next = L->front;
    	L->length++;
    	if(L->front == NULL){
            L->back = n;

    	}else{
    		L->front->previous = n;
    	}
    	L->front = n;
        if(L->index >= 0) L->index++;
    }

    void append(List L, int data){
        Node n = malloc(sizeof(struct NodeObj));
        n->value = data;
        n->previous = L->back;
        n->next = NULL;
        L->length++;

        if(L->back == NULL){
            L->front = n;
        }else{
            L->back->next = n;
        }
        L->back = n;
    }

    // void insertBefore(List L, int data){
    // 	Node n = malloc(sizeof(struct NodeObj));
    // 	L->length++;
    // 	n->next = L->cursor;
    //     n->previous = L->cursor->previous;
    //     n->value = data;
    // 	if(L->length >0 && L->index >=0){
    // 		if(L->cursor->previous != NULL){
    // 			L->cursor->previous->next=n;
    // 		}else{
    // 			L->front=n;
    // 		}
    // 		L->cursor->previous = n;
    // 	}
    // }

    void insertBefore(List L, int data){
    Node n = malloc(sizeof(struct NodeObj));
    n->value = data;
    n->previous = L->cursor->previous;
    n->next = L->cursor;
    if(n->previous != NULL && L->cursor->next != NULL){
        L->cursor->previous->next = n;
    }else{
        L->front = n;
    }
    L->cursor->previous = n;
    L->length++;
    L->index++;
}


    void insertAfter(List L, int data){
    	Node n = malloc(sizeof(struct NodeObj));
    	n->value = data;
    	n->previous = L->cursor;
    	n->next = L->cursor->next;
    		if(n->previous != NULL && L->cursor->next != NULL){
    			L->cursor->next->previous = n;
    		}else{
    			L->back = n;
    		}
    		L->cursor->next = n;
            L->length++;
    }

    void deleteFront(List L){
        if(L->length>0){
            if(L->index == 0){
              L->cursor = NULL;
              L->index = -1;
          }
          if(L->back == L->front) L->back = NULL;
          Node x = L->front;
          L->front = L->front->next;
          free(x);
          L->front->previous = NULL;
          if(L->index >= 0) L->index--;
          L->length--;
        }
    }    

    // void deleteBack(List L){
    // 	if(L->length>0){
    // 		L->back = L->back->previous;
    // 		L->back->previous = NULL;
    // 	}
    // }

    void deleteBack(List L){
      if(L==NULL){
        printf("Null pointer error");
        return;
    }
    if(L->length>0){
        if(L->index == L->length-1){
          L->cursor = NULL;
          L->index = -1;
      }
      if(L->back == L->front) L->front = NULL;
      Node tempNode = L->back;
      L->back = L->back->previous;
      free(tempNode);
      if(L->back)L->back->next = NULL;
      L->length--;
  }
} 

    void delete(List L){
    	if(L->length > 0 && L->index >=0){
    		L->length--;
    		L->cursor->value = -1;
    		L->cursor->next->previous = L->cursor->previous;
    		L->cursor->previous->next = L->cursor->next;
    	}
    	if(L->cursor == L->front){
    		deleteFront(L);
    	}
    	if(L->cursor == L->back){
    		deleteBack(L);
    	}
    }

    // void printList(FILE* out, List L){
    //   Node N = L->front;
    //   while(N != NULL){
    //     printf("printList err\n");
    //     fprintf(out, "%d ", N->value);
    //     N = N->next;
    //   }
    // }

    void printList (FILE *out, List L){
      if (L == NULL){
        printf("List.c: 352 passing null List ptr to printList()");
        return;
    }
    if (out == NULL){
        printf("List.c: 355: passing null FILE ptr to printList()");
        return;
    }

    Node temp = L->front;
    while (temp != L->back){
        fprintf(out, "%d ", temp->value);
        temp = temp->next;
    }
    if(temp != NULL) fprintf(out, "%d", temp->value);

}


    List copyList(List L){
        List x = newList();
        Node N = L->front;
        while(N != NULL){
            append(x, N->value);
            N = N->next;
        }
        return x;
    }

void freeList (List* pL){
      if ((*pL) == NULL) {
        printf("List.c: 43: Passing Nullpointer to freeList()");
        return;
    }
    clear(*pL);
    free(*pL);
    *pL = NULL;
}

void clear(List L){
  if (L == NULL){
    printf("List.c : 130: Passing null ptr to clear()");
    return;
}
Node front = L->front;
while (front != NULL){
    Node temp = front->next;
    free(front);
    front = temp;
}
L->front = NULL;
L->back = NULL;
L->cursor = NULL;
L->length = 0;
L->index = -1;
}

    List concat(List L, List L2){
        List T = copyList(L);
        Node N = L2->front;
        while(N != NULL){
            append(T, N->value);
            N = N->next;
        }
        return T;
    }





    

File: code/pa3/List.java
class List{
        private class node{
                Object value;
                node next;
                node previous;


                node(Object x){
                        value = x;
                        next = null;
                        previous = null;
                }
                
            
                public String toString(){
                    return "" + value;
        }

        }

        private node cursor;
        private node back;
        private node front;
        private int length;
        private int index;

        public List(){
                cursor=null;
                front=null;
                back=null;
                length=0;
                //index=-1;
        }


        public int index(){
                if(cursor==null){
                        return -1;
                }
                return index;
        }

        public int length(){
                return length;
        }

        public Object front(){
                if(length()>0){
                        return front.value;
                }
                throw new RuntimeException(
                        "Queue Error: front() called on empty Queue");
        }

        public Object back(){
                if(length()>0){
                        return back.value;
                }
                throw new RuntimeException(
                        "Queue Error: back() called on empty Queue");
        }

        public Object get(){
                if(length>0 && index()>=0){
                        return cursor.value;
                }
                return -1;
        }

        public boolean equals(Object L){
                node x1 = front;
                node x2 = ((List)L).front;
                while(x1!=null && x2!=null){
                        if(x1.value != x2.value){
                                return false;
                        }
                        x1=x1.next;
                        x2=x2.next;
                }
                return true;
        }

        public void clear(){
                cursor=null;
                front=null;
                back=null;
                length=0;
                index=-1;
        }

        public void moveFront(){
                if(front != null){
                        cursor = front;
                        index = 0;
                }
        }

        public void moveBack(){
                if(back!=null){
                        cursor = back;
                        index = length-1; //n-1//
                }
        }

        public void movePrev(){
                if(index()>0 && index()<= length()-1){
                        cursor = cursor.previous;
                }else{
                        cursor=null;
                }
        }

        /*public void moveNext(){
                int index = index();
                if(index >= 0 && index < length() - 1){
                    cursor = cursor.next;
            } else {
                    cursor = null;
                    System.out.println("moveNext null");

            }
    }*/

void moveNext(){
        if(cursor != null){
                if(index <=(length())-1){
                        cursor = cursor.next;
                        index++;
                }
                if(index == length()){
                        cursor =null;
                        index = -1;
                }
        }
}


    /*void moveNext()
    {
          if (cursor==back) cursor = null;
          if (cursor!=null) cursor = cursor.next;
  }*/

        /*public void moveNext(){
                if(cursor!=back){
                        cursor=cursor.next;
                        index++;
                }else{
                        index=-1;
                        cursor=null;
                }
        }*/

        /*public int getSize(){
                return matrix.length;
        }*/

        public void prepend(Object x){
                node n = new node(x);
                index++;
                length++;
                if(front!=null){
                        front.previous = n;
                        n.next = front;
                        front = n;
                }else{
                        back=n;
                }
                front=n;
        }

       /* public void append(Object x){
                node n = new node(x);
                index++;
                if(back!=null){
                        back.next=n;
                        n.previous=back;
                }else{
                        front = n;
                }
                back = n;
        }*/

        public void append(Object data){
        length++;
        index++;
        node temp = new node(data);
        if (back != null) {
            back.next = temp;
        } else {
            front = temp;
        }
        back = temp;
    }

        public void insertBefore(Object x){
                node n = new node(x);
                length++;
                n.next=cursor;
                if(length() >0 && index() >=0){
                        if(cursor.previous != null){
                                cursor.previous.next=n;
                        }else{
                                front = n;
                        }
                        cursor.previous = n;
                }
        }

        public void insertAfter(Object x){ //
                node n = new node(x);
                n.previous=cursor;
                n.next=cursor.next;
                if(length() > 0 && index() >=0){
                        length++;
                        if(cursor.next != null){
                                cursor.next.previous=n;
                        }else{
                                back=n;
                        }
                        cursor.next=n;
                }
        }

        public void deleteFront(){
                if(length()>0){
                        front=front.next;
                        if(front != null){
                                front.previous=null;
                        }
                        index--;

                        //length--;
                }
        }

        /*public void deleteBack(){
                if(length()>0){
                        back=back.previous;
                        back.previous=null;
                        index--;
                        //length--;
                }
        }*/

        public void deleteBack(){
                if (length() > 0){
                    length--;
                    if(cursor == back) cursor = null;
                    back = back.previous;
                    if(back != null) back.next = null;

                }
        }

/*        void delete() throws RuntimeException 
        {
          if(length()<1)
          {
               throw new RuntimeException(
                    "List Error: insertAfter() called on empty List");
       } 
       if(index()<0)
       {
               throw new RuntimeException(
                    "Cursor Error: insertAfter() called on null Index");
       }
       if(cursor==front) deleteFront();
       if(cursor==back) deleteBack();
       cursor.next.previous = cursor.previous;
       cursor.previous.next = cursor.next;
       cursor = null;
       length--;
}*/


        public void delete(){
                if(length()>0 && index()>=0){
                        length--;
                        cursor.value=-1;
                        cursor.next.previous = cursor.previous;
                        cursor.previous.next = cursor.next;
                }
                if(cursor==front){
                        deleteFront();
                }
                if(cursor==back){
                        deleteBack();
                }
        }

        public String toString(){
                node n = front;
                String s1="";
                while(n!=null){
                        s1=s1+n+" ";
                        n=n.next;
                }
                return s1;
        }

        /*List copy (){
                List newList= new List();
                node n = front;
                while(n!=null){
                        newList.append(n.value);
                        n=n.next;
                }
                return newList;
        }

        List concat(List x){
                List newList = copy();
                node n = x.front;
                while(n!=null){
                        newList.append(n.value);
                        n=n.next;
                }
                return newList;
        }*/
}
